xinference/model/llm/vllm/core.py

```diff
@@ -319,16 +323,42 @@ async def async_generate(
 
         async def stream_results() -> AsyncGenerator[CompletionChunk, None]:
             previous_texts = [""] * sanitized_generate_config["n"]
+            tools_token_filter = ChatModelMixin._tools_token_filter(self.model_family)
             async for _request_output in results_generator:
                 chunk = self._convert_request_output_to_completion_chunk(
                     request_id=request_id,
                     model=self.model_uid,
                     request_output=_request_output,
                 )
+
                 for i, choice in enumerate(chunk["choices"]):
                     delta = choice["text"][len(previous_texts[i]) :]
                     previous_texts[i] = choice["text"]
                     choice["text"] = delta
+
+                if tools:
+                    # only handle the first choice
+                    choice = chunk["choices"][0]
+                    if choice["finish_reason"] is not None:
```

Comment:
Why use `choice["finish_reason"] is not None` as the condition?

Reply:
Or choice["finish_reason"] == "stop"? Anyway we need a way to check if the model wants to call the tool, at least for Qwen `choice["finish_reason"] is not None` is feasible.

Reply:
The `finish_reason` comes from the vllm output here. Did you make sure this  reliable?